tests/test_interpolation.py

```diff
@@ -753,3 +755,244 @@ def test_none_value_in_quoted_string(restore_resolvers: Any) -> None:
+        ),
+    ],
+)
+def test_interpolation_type_validated_error(
+    cfg: Any,
+    key: str,
+    expected_error: Any,
+    common_resolvers: Any,
+) -> None:
+    cfg = OmegaConf.structured(cfg)
+
+    with expected_error:
+        cfg[key]
+
+    assert OmegaConf.select(cfg, key, throw_on_resolution_failure=False) is None
+
+
+@pytest.mark.parametrize(
+    ("cfg", "key"),
+    [
+        pytest.param({"dict": "${identity:{a: 0, b: 1}}"}, "dict.a", id="dict"),
+        pytest.param(
+            {"dict": "${identity:{a: 0, b: {c: 1}}}"},
+            "dict.b.c",
+            id="dict_nested",
+        ),
+        pytest.param({"list": "${identity:[0, 1]}"}, "list.0", id="list"),
+        pytest.param({"list": "${identity:[0, [1, 2]]}"}, "list.1.1", id="list_nested"),
+    ],
+)
+def test_interpolation_readonly_resolver_output(
+    common_resolvers: Any, cfg: Any, key: str
+) -> None:
+    cfg = OmegaConf.create(cfg)
+    sub_key: Any
+    parent_key, sub_key = key.rsplit(".", 1)
+    try:
+        sub_key = int(sub_key)  # convert list index to integer
+    except ValueError:
+        pass
+    parent_node = OmegaConf.select(cfg, parent_key)
+    assert parent_node._get_flag("readonly")
+
+
+def test_interpolation_readonly_node() -> None:
+    cfg = OmegaConf.structured(User(name="7", age=II("name")))
+    resolved = cfg._get_node("age")._dereference_node()
+    assert resolved == 7
+    # The `resolved` node must be read-only because `age` is an integer, so the
+    # interpolation cannot return directly the `name` node.
+    assert resolved._get_flag("readonly")
+
+
+def test_type_validation_error_no_throw() -> None:
+    cfg = OmegaConf.structured(User(name="Bond", age=SI("${name}")))
+    bad_node = cfg._get_node("age")
+    assert bad_node._dereference_node(throw_on_resolution_failure=False) is None
+
+
+@pytest.mark.parametrize(
+    ("cfg", "expected"),
+    [
+        ({"a": 0, "b": 1}, {"a": 0, "b": 1}),
+        ({"a": "${y}"}, {"a": -1}),
+        ({"a": 0, "b": "${x.a}"}, {"a": 0, "b": 0}),
+        ({"a": 0, "b": "${.a}"}, {"a": 0, "b": 0}),
+        ({"a": "${..y}"}, {"a": -1}),
+    ],
+)
+def test_resolver_output_dict_to_dictconfig(
+    restore_resolvers: Any, cfg: Dict[str, Any], expected: Dict[str, Any]
+) -> None:
+    OmegaConf.register_new_resolver("dict", lambda: cfg)
+    c = OmegaConf.create({"x": "${dict:}", "y": -1})
+    assert isinstance(c.x, DictConfig)
+    assert c.x == expected
+
+
+@pytest.mark.parametrize(
+    ("cfg", "expected"),
+    [
+        ([0, 1], [0, 1]),
+        (["${y}"], [-1]),
+        ([0, "${x.0}"], [0, 0]),
+        ([0, "${.0}"], [0, 0]),
+        (["${..y}"], [-1]),
+    ],
+)
+def test_resolver_output_list_to_listconfig(
+    restore_resolvers: Any, cfg: List[Any], expected: List[Any]
+) -> None:
+    OmegaConf.register_new_resolver("list", lambda: cfg)
+    c = OmegaConf.create({"x": "${list:}", "y": -1})
+    assert isinstance(c.x, ListConfig)
+    assert c.x == expected
```

Comment:
You can also add a quick check to confirm those are readonly (here and in dict).
```suggestion
    assert c.x == expected
    assert x._get_node("x")._get_flag("readonly")
```

Reply:
Done in 8245769
I ended up adding a utility function since it can't be a one-liner because of mypy